### ISSUE TITLE
[BUG] Invalid Logins are not sent to log, because of this it is not possible to create a fail2ban jail or crowdsec parser.

### ISSUE THREAD

**Issue by Jieiku** [osTicket/osTicket#5716]
### Prerequisites

* [yes ] Can you reproduce the problem in a fresh installation of the "develop" branch? yes
* [no ] Do you have any errors in the PHP error log, or javascript console?
* [yes ] Did you check the [osTicket forums](https://forum.osticket.com/)?
* [yes ] Did you [perform a cursory search](https://github.com/osTicket/osTicket/issues) to see if your bug or enhancement is already reported?

### Description

It seems invalid login attempts with the IP address are not sent to the log file.

### Steps to Reproduce

sudo tail -f /var/log/syslog

Now try and log in to osticket with invalid login name and password
It never gets sent to the log

**Expected behavior:** [What you expected to happen]

I expect that when an invalid login attempt happens it should get sent to the log file.

**Actual behavior:** [What actually happened]

Unfortunately it never got sent to the log.

### Versions

Admin panel -> Dashboard -> Information which also additionally gives you information about your server.

Ubuntu 20.04
osTicket Version | v1.14.2 (cba6035)
Web Server Software | Apache/2.4.41 (Ubuntu)
MySQL Version | 8.0.22
PHP Version | 7.4.3

### Notes

I do see that on the admin panel dashboard there is log info for invalid login attempts, but as far as I can tell these do not exist in any log files, so I am assuming they must be within the database, this prevents me from being able to use the fail2ban program. I have custom fail2ban actions, such as notifying a central server so that the offending IP is banned at the firewall at the edge of the network directly.

As far as I can tell its as easy as adding this snippet to your php page that handles the login process:

```
$ipaddress = getenv("REMOTE_ADDR");
openlog("phperrors", LOG_PID | LOG_PERROR);
syslog(LOG_ERR, "Wrong username or password from " . $ipaddress);
closelog();
```

Hope this can be resolved, if the log file does exist and I somehow missed it then I would appreciate some help finding it.

**Comment by ntozier** [comment]
I would be very surprised if any web application could write directly to syslog.  Things like this are usually handled at the webserver level.

for Apache see
https://httpd.apache.org/docs/2.4/logs.html

You would also need to go to **Admin panel -> Settings -> System** and set *Default Log Level* to ERROR.
And it will log things to osTicket's internal syslog like this:
```
log_id;log_type;title;log;logger;ip_address;created;updated
#;Error;Excessive login attempts;Excessive login attempts by a user. Username: <username here> IP:<IPA.DRE.SSh.ere> Time Nov 27, 2020, 3:05 pm UTC Attempts: 5;;<IPA.DRE.SSh.ere>;2020-11-27 15:05:14;2020-11-27 15:05:14;```

note: it does not log every single time a user fails to log in.  
You also have control over how much is considered excessive for Agents:
**Admin panel -> Settings -> Agents**
and Users:
**Admin panel -> Settings -> Users**

**Comment by Jieiku** [comment]
I am a sysadmin, plenty of applications write to the syslog without issue, and I have fail2ban configured for all of them, to list a few: nextcloud, wordpress, freepbx, nagios, gitea, emoncms, ...

In fact most apps can either write to syslog or to some form of error.log, security.log, or application.log etc.

I actually dont care what file the log info is sent to, so long as it is an external file(not only in some database) that fail2ban can monitor.

It also does not need to write to the log file by default, a config option could be set such as
`securitylog=true;`

**Comment by JediKev** [comment]
@xekon 

In the meantime you could write a very simple script to pull the logs from the db and export to a file of your choice.

Cheers.

**Comment by Jieiku** [comment]
@JediKev  Thanks for the idea, I would probably just modify the php login page osticket and add this snippet:

```
$ipaddress = getenv("REMOTE_ADDR");
openlog("phperrors", LOG_PID | LOG_PERROR);
syslog(LOG_ERR, "Wrong username or password from " . $ipaddress);
closelog();
```
but I appreciate the response. In the meantime I do not have osticket internet facing, it is internal use only because of not being able to use fail2ban.

**Comment by Jieiku** [comment]
@ntozier  I set settings>system>default log level to ERROR, I next set Agents and Users to agent excessive logins 1 failed login attempts and 1 minutes locked out.

I then try logging in repeatedly with the wrong password while watching both the error and access log:
```
sudo grc tail -f /var/log/apache2/access.log
sudo grc tail -f /var/log/apache2/error.log
```

Every time I hit the page with an invalid login I can see a new line in the access log, but invalid login attempts never get sent to either the error or access logs.

**Comment by protich** [comment]
Yes, it's possible to log stuff  e.g php error log via `error_log()` but osTicket doesn't do so for any type of errors - so this is not a bug but rather a feature request. I suggest a plugin `log-error` that would listen to error log signal and do whatever it wants to it  e.g log it.

If anyone wants to take a stab at it then I'll be happy to provide some guidance.

Thanks,

**Comment by Jieiku** [comment]
I would like to try making the plugin. is there an example plugin framework I can get started with?

**Comment by ntozier** [comment]
I maintain a list of resources for osTicket on my web site [tmib.net/resources-for-osticket](https://tmib.net/resources-for-osticket/) and I have this list cross posted on the forums at https://forum.osticket.com/d/92286-resources-for-osticket.  I think that it should go with out saying that there is also a pretty active community at the forums who may be able to assist with direct questions as well.

Here is the Development Resources section of that list.
[Repo – Signals Docs](https://github.com/osTicket/osTicket/blob/develop/setup/doc/signals.md) – How signals work.
[Unofficial Plugin Development Guide](https://github.com/poctob/OSTEquipmentPlugin/wiki/Plugin-Development-Introduction) – Provides reverse-engineered instructions on how to develop an osTicket plugin.

**Comment by Jieiku** [comment]
Honestly I could not really find any resources to get started with that gave a clear example such as a "Hello World" plugin, something bare bones and easy to understand.

I did however resolve this and have fail2ban working.

`sudo nano /var/www/osTicket/scp/login.php`

find this line:
`$show_reset = true;`

below it add these lines:
```
    $ipaddress = getenv("REMOTE_ADDR");
    openlog("phperrors", LOG_PID | LOG_PERROR, LOG_SYSLOG);
    syslog(LOG_ERR, "Bad Login from " . $ipaddress);
    closelog();
```

now watch your log file while trying to log into admin panel with incorrect password:
`sudo tail -f /var/log/apache2/error.log`

you will see the output like this:

`phperrors[280]: Bad Login from 111.222.33.44`

To create a fail2ban jail for this:

`sudo nano /etc/fail2ban/filter.d/osticket.conf`
```
[Definition]
failregex = .*Bad Login from <HOST>$
ignoreregex =
```

`sudo nano /etc/fail2ban/jail.local`
```
[osticket]
enabled = true
filter = osticket
action = iptables-allports
logpath = /var/log/apache2/error.log
maxretry = 1
```

now reboot:
`sudo reboot`

Congratulations, you are now banning the people trying to brute their way into your admin panel.

**Comment by conandrum** [comment]
Hi @Jieiku, @protich @JediKev (@ntozier is probably already informed)

Just FYI, a new plugin was recently developed for this reason.

**authLoginFailedLog**
osTicket does not log anything to any external log file. Instead it uses its own database for logging. This is problematic when you need to protect osTicket against brute-force attacks on the login pages.
This plugin allows osTicket to log failed authentication attempts in the apache error log. As a result you can now use utilities such as fail2ban to query the log file for such and react in realtime to protect osTicket.

**authLoginFailedLog is a free plugin for Motherload for osTicket.**
It is available in the download section on the product page here:
https://www.cartmega.com/software/osticket/motherload-for-osticket.html

The beauty of Motherload is that you can quickly write such small useful plugins without having to bear the overhead of creating a full blown osTicket plugin with all its bells and whistles.
I hope this helps some of you. 😃

**Comment by breezecomputersolutions** [comment]
**Nice!** Works great. Thank you for taking the time to write this up!

**Comment by Jieiku** [comment]
I need to revisit this later, we have been moving from fail2ban to crowdsec, but crowdsec has the same problem of needing log files to work with, which osTicket does not create.

My solution should work for crowdsec as well, but I honestly hacked it together pretty quickly, happy to hear it is still working.

**Comment by breezecomputersolutions** [comment]
great. thanks for letting me know about crowdsec!

**Comment by conandrum** [comment]
My previous message (Feb 13) was deleted for some reason, so I am chiming in again.

In order to log your failed-login-attempts into a file, in osTicket , you can write a plugin (no core file modifications necessary).
Basically you need to **listen for the auth.login.failed signal** and then write to the apache log file with **openlog/syslog/closelog commands**. 

I wrote a short PHP script called [**AuthLoginFailedLog**](https://www.cartmega.com/motherload-for-osticket.html) that does this and is free to download. 
It needs osTicket Motherload  to run though, so you can use it as a template for your own plugin if you like. 

**Bottom line: do not modify your core files... write a short plugin instead. Feel free to use my code as your starting point.**

**Comment by breezecomputersolutions** [comment]
Great. I will check it out. I appreciate your help!

**Comment by Jieiku** [comment]
@conandrum I did not delete your previous message, but if I were the admin I might have. My solution works without paying for something like motherload. It should not be necessary to pay for something simply to log a message into a log.

**Comment by conandrum** [comment]
@Jieiku No doubt that your solution works for many people. Others may opt to put it into a plugin so that it keeps working through osTicket upgrades. Others still, may opt to get something off the shelf and be done with it. 
Having options is never a bad thing.

**Comment by ntozier** [comment]
So if I needed to do something like this I would make a script and run it via cron every minute or few minutes.

You could probably use something like this:
```SELECT * FROM ost_syslog where created > now() - INTERVAL 2 MINUTE INTO OUTFILE '\path\to\logname.txt'```

Then you would have a log file for fail2ban

**Comment by breezecomputersolutions** [comment]
> **Nice!** Works great. Thank you for taking the time to write this up!

Update: I had to remove the code because even valid logins were being logged as "Bad Login..."

**Comment by a1ad** [comment]
Hi, I am also interested in bad auth logs to feed a Crowdsec parser. The cleanest way to do so is to create a plugin. Or better yet, an osTicket PR.
